exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java

```diff
@@ -491,6 +556,25 @@ private void inferOutputFieldsFromLeftSide() {
       }
     }
 
+    private void inferOutputFieldsFromRightSide() {
```

Comment:
It seems like the inferencing is repeated a few times.  You could have a single function like inferOutputFieldNameAndType(outputFieldNames,  batch).  The field type is always inferred from the supplied batch.  For the output names, if the outputFieldNames is null, then use the output field names as well as type from that batch (this would be the case for left input).  If not null, then use the field names.  Would that work ?